jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/Invocable.java

```diff
@@ -241,4 +263,156 @@ public InvocationType getInvocationType()
         }
         return result;
     }
+
+    /**
+     * An extension of {@link java.util.concurrent.CompletableFuture} that is an {@link Invocable}.
+     * The {@link InvocationType} is the type passed in construction (default NON_BLOCKING).
+     * Methods on {@link java.util.concurrent.CompletableFuture} that may act in contradiction to the passed
+     * {@link InvocationType} are extended to throw {@link IllegalStateException} in those circumstances.
+     * <p>
+     * Counterintuitively, if the blocking APIs like {@link #get()} are to be used, the {@link InvocationType}
+     * should be {@link InvocationType#NON_BLOCKING}, as the wake-up callbacks used will not block.
+     *
+     * @param <V> The type of the result
+     * @deprecated This class in only used for deprecated usages of CompletableFuture
+     */
+    @Deprecated(forRemoval = true, since = "12.0.15")
+    class InvocableCompletableFuture<V> extends java.util.concurrent.CompletableFuture<V> implements Invocable
```

Comment:
I don't think we should add new APIs, throwing ISEs from CompletableFuture methods that don't normally throw, it breaks least surprise.
I'd just deprecate the relevant APIs.

Reply:
I think we should not care about least surprise here, as we do not want people to use this API at all.
They are obtaining a CF after passing arguments to say how they will using it... if they end up using it wrongly (and risking deadlock) then we can catch that case with an ISE.    It is a protective layer whilst we remove this usage entirely.

Reply:
We just deprecate and document it. 
It is not solving the issue, just throwing an exception in the user's face, which will likely break the application anyway, so it does not really protect users.

Reply:
It protects users by throwing during development and testing rather than waiting until the last thread in production deadlocks.

Reply:
As discussed in hangout, we will keep this whilst developing this PR and then consider removing it (note it has already helped me find one coding error)